python/ray/serve/_private/application_state.py

```diff
@@ -141,127 +229,191 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
-                    break
-            if self._deployments_to_delete:
-                mark_delete = False
-            self._ready_to_be_deleted = mark_delete
-            self._process_terminating_deployments()
+        if not self._deploy_obj_ref:
             return
 
-        if self._status == ApplicationStatus.DEPLOYING:
-            if self._deploy_obj_ref:
-                finished, pending = ray.wait([self._deploy_obj_ref], timeout=0)
-                if pending:
-                    return
-                self._deploy_obj_ref = None
-                try:
-                    ray.get(finished[0])
-                    logger.info(f"Deploy task for app '{self._name}' ran successfully.")
-                except RayTaskError as e:
-                    # NOTE(zcin): we should use str(e) instead of traceback.format_exc()
-                    # here because the full details of the error is not displayed
-                    # properly with traceback.format_exc(). RayTaskError has its own
-                    # custom __str__ function.
-                    self._update_status(
-                        ApplicationStatus.DEPLOY_FAILED,
-                        status_msg=f"Deploying app '{self._name}' failed:\n{str(e)}",
-                    )
-                    logger.warning(self._status_msg)
-                    return
-                except RuntimeEnvSetupError:
-                    self._update_status(
-                        ApplicationStatus.DEPLOY_FAILED,
-                        status_msg=(
-                            f"Runtime env setup for app '{self._name}' "
-                            f"failed:\n{traceback.format_exc()}"
-                        ),
-                    )
-                    logger.warning(self._status_msg)
-                    return
-                except Exception:
-                    self._update_status(
-                        ApplicationStatus.DEPLOY_FAILED,
-                        status_msg=(
-                            "Unexpected error occured while deploying "
-                            f"application '{self._name}':"
-                            f"\n{traceback.format_exc()}"
-                        ),
-                    )
-                    logger.warning(self._status_msg)
-                    return
-            deployments_statuses = (
-                self._deployment_state_manager.get_deployment_statuses(self.deployments)
-            )
-            num_health_deployments = 0
-            unhealthy_deployment_names = []
-            for deployment_status in deployments_statuses:
-                if deployment_status.status == DeploymentStatus.UNHEALTHY:
-                    unhealthy_deployment_names.append(deployment_status.name)
-                if deployment_status.status == DeploymentStatus.HEALTHY:
-                    num_health_deployments += 1
-
-            if len(unhealthy_deployment_names) != 0:
+        if self._check_obj_ref_ready(self._deploy_obj_ref):
```

Comment:
Question: Should there be a time out for this just in case if deploy obj ref hangs forever?

Reply:
This function doesn't block on the object ref, it uses timeout=0. If the deploy obj ref hangs forever, _check_obj_ref_ready will continuously return false, and the application will just be stuck in DEPLOYING forever. We currently don't have anything that checks on how long an app has been deploying for - we could potentially add logic that moves it to unhealthy, but I think if the app is stuck on deploying it also is enough signal to the user that something is wrong.